internal/dbutil/sqliteutil/query.go

```diff
@@ -0,0 +1,186 @@
+	indexDefinitions := make([]*Definition, 0)
+
+	// find tables and indexes
+	err := func() error {
+		rows, err := db.Query(`
+			SELECT name, type, sql FROM sqlite_master WHERE sql NOT NULL AND name NOT LIKE 'sqlite_%'
+		`)
+		if err != nil {
+			return err
+		}
+		defer func() { err = errs.Combine(err, rows.Close()) }()
+
+		for rows.Next() {
+			var defName, defType, defSQL string
+			err := rows.Scan(&defName, &defType, &defSQL)
+			if err != nil {
+				return err
+			}
+			if defType == "table" {
+				tableDefinitions = append(tableDefinitions, &Definition{name: defName, sql: defSQL})
+			} else if defType == "index" {
+				indexDefinitions = append(indexDefinitions, &Definition{name: defName, sql: defSQL})
+			}
+		}
+
+		return rows.Err()
+	}()
+	if err != nil {
+		return nil, err
+	}
+
+	// discover tables
+	err = func() (err error) {
+		for _, definition := range tableDefinitions {
+			table := schema.EnsureTable(definition.name)
+
+			tableRows, err := db.Query(`PRAGMA table_info(` + definition.name + `)`)
+			if err != nil {
+				return err
+			}
+			defer func() { err = errs.Combine(err, tableRows.Close()) }()
+
+			for tableRows.Next() {
+				var defaultValue sql.NullString
+				var index, name, columnType string
+				var pk int
+				var notNull bool
+				err := tableRows.Scan(&index, &name, &columnType, &notNull, &defaultValue, &pk)
+				if err != nil {
+					return err
+				}
+
+				column := &dbschema.Column{
+					Name:       name,
+					Type:       columnType,
+					IsNullable: !notNull && pk == 0,
+				}
+				table.AddColumn(column)
+				if pk > 0 {
+					if table.PrimaryKey == nil {
+						table.PrimaryKey = make([]string, 0)
+					}
+					table.PrimaryKey = append(table.PrimaryKey, name)
+				}
+
+			}
+
+			matches := rxUnique.FindAllStringSubmatch(definition.sql, -1)
+			for _, match := range matches {
+				// TODO feel this can be done easier
+				var columns []string
+				for _, name := range strings.Split(match[1], ",") {
+					columns = append(columns, strings.TrimSpace(name))
+				}
+
+				table.Unique = append(table.Unique, columns)
+			}
+
+			keysRows, err := db.Query(`PRAGMA foreign_key_list(` + definition.name + `)`)
+			if err != nil {
+				return err
+			}
+			defer func() { err = errs.Combine(err, keysRows.Close()) }()
+
+			for keysRows.Next() {
+				var id, sec int
+				var tableName, from, to, onupdate, ondelete, match string
+				err := keysRows.Scan(&id, &sec, &tableName, &from, &to, &onupdate, &ondelete, &match)
+				if err != nil {
+					return err
+				}
+
+				column, found := table.FindColumn(from)
+				if found {
+					if ondelete == "NO ACTION" {
+						ondelete = ""
+					}
+					if onupdate == "NO ACTION" {
+						onupdate = ""
+					}
+					column.Reference = &dbschema.Reference{
+						Table:    tableName,
+						Column:   to,
+						OnUpdate: onupdate,
+						OnDelete: ondelete,
+					}
+				}
+			}
+		}
+		return err
+	}()
+	if err != nil {
+		return nil, err
+	}
+
+	// discover indexes
```

Comment:
it might be worth breaking the "find tables and indexes", "discover tables" and "discover indexes" code into separate helper functions for readability but this isn't critical.